runtime/alloc/arena.cpp

```diff
@@ -8,16 +8,10 @@
 #include "runtime/header.h"
 #include "runtime/alloc.h"
  
-typedef struct {
-  char* next_block;
-  char* next_superblock;
-  char semispace;
-} memory_block_header;
-
 const size_t BLOCK_SIZE = 1024 * 1024;
 
 #define mem_block_header(ptr) \
-  ((memory_block_header *)(((uintptr_t)(ptr)) & ~(BLOCK_SIZE-1)))
+  ((memory_block_header *)(((uintptr_t)(ptr) - 1) & ~(BLOCK_SIZE-1)))
```

Comment:
I am not sure I get the fix here. Can you explain?

Reply:
A pointer pointing to an address that is megabyte aligned might point one past the last object in the block, but it will never point at the start of an object because that space within the block is taken up by the memory block header. However, if you are attempting to compute the block header of a pointer one past the end of the block, the old code would instead give you the block header of the next block sequentially in memory. This causes problems when trying to compare whether two scan pointers are members of the same block or not. Because we just chop off the low order bits, and because the first allocated byte in a block is always 24 bytes after the start of the block, we can make sure that this macro always takes you to the address associated with the block header of the block the pointer points to or is one past the end of by simply subtracting one prior to chopping off the low order bits.